aws/service.ts

```diff
@@ -247,74 +247,109 @@ function computeImage(
-                //
-                // See: http://docs.aws.amazon.com/cli/latest/reference/ecr/get-authorization-token.html
-                if (!registryId) {
-                    throw new RunError("Expected registry ID to be defined during push");
-                }
-                const credentials = await aws.ecr.getCredentials({ registryId: registryId });
-                const decodedCredentials = Buffer.from(credentials.authorizationToken, "base64").toString();
-                const [username, password] = decodedCredentials.split(":");
-                if (!password || !username) {
-                    throw new RunError("Invalid credentials");
-                }
-                return {
-                    registry: credentials.proxyEndpoint,
-                    username: username,
-                    password: password,
-                };
-            });
-            if (imageName) {
-                buildImageCache.set(imageName, imageDigest);
-            }
-            imageDigest.apply(d =>
-                pulumi.log.debug(`    build complete: ${imageName} (${d})`, repository));
-        }
-
-        preEnv.IMAGE_DIGEST = imageDigest;
-
-        return pulumi.all([repository.repositoryUrl, pulumi.all(preEnv)])
-                     .apply(([url, e]) => ({ image: url, environment: e }));
+        return computeImageFromBuild(preEnv, container.build, imageName, repository);
     }
     else if (container.image) {
-        return pulumi.all(preEnv).apply(e => ({ image: imageName, environment: e }));
+        return computeImageFromImage(preEnv, imageName);
     }
     else if (container.function) {
-        const func = container.function;
-        preEnv.PULUMI_SRC = pulumi.runtime.serializeFunctionAsync(func);
-
-        // TODO[pulumi/pulumi-cloud#85]: Put this in a real Pulumi-owned Docker image.
-        // TODO[pulumi/pulumi-cloud#86]: Pass the full local zipped folder through to the container (via S3?)
-        return pulumi.all(preEnv).apply(e => ({ image: imageName, environment: e }));
+        return computeImageFromFunction(container.function, preEnv, imageName);
     }
     else {
         throw new RunError("Invalid container definition: `image`, `build`, or `function` must be provided");
     }
 }
 
+function computeImageFromBuild(
+        preEnv: Record<string, pulumi.Input<string>>,
+        build: string | cloud.ContainerBuild,
+        imageName: string,
+        repository: aws.ecr.Repository): pulumi.Output<ImageOptions> {
+
+    // This is a container to build; produce a name, either user-specified or auto-computed.
+    pulumi.log.debug(`Building container image at '${build}'`, repository);
+    const { repositoryUrl, registryId } = repository;
+
+    return pulumi.all([repositoryUrl, registryId]).apply(([repoUrl, regId]) =>
+        computeImageFromBuildWorker(preEnv, build, imageName, repoUrl, regId, repository));
```

Comment:
Core pattern: if you depend on outputs, unwrap them early and pass them along.  that way you don't have to figure out how to do it in the middle of some complex logic flow.

Reply:
Note: i coudl have made computeImageFromBuildWorker an inner function (which captured outer values).  but i personally found that more confusing, and i liked that there were two distinct functions that were not sharing.

Reply:
@lukehoban Rgd `pulumi.all([repositoryUrl, registryId]).apply(([repoUrl, regId])`

I've realized our linter rule about shadowing variables actually hurts us with this pattern.  Ideally you really jsut want to say:

`pulumi.all([repositoryUrl, registryId]).apply(([repositoryUrl, registryId])`

You don't really want to have to think up new names.  And here the shadowing is actually *good* since you want these inner parameters to mask that outer variables.  

Thoughts on removing this linting rule and cleaning this up across all our code?